File: typescript/src/project.ts
import { Instance, InstanceData } from "./instance";
import { MainContractAddress, SkaleABIFile } from "./domain/types";
import axios, { HttpStatusCode } from "axios";
import { InstanceNotFound } from "./domain/errors/instance/instanceNotFound";
import { ListedNetwork } from "./listedNetwork";
import { Network } from "./network";
import {
    NetworkNotFoundError
} from "./domain/errors/network/networkNotFoundError";
import { ProjectMetadata } from "./metadata";
import { REPOSITORY_URL } from "./domain/constants";
import { ethers } from "ethers";

export abstract class Project<ContractType> {
    protected metadata: ProjectMetadata;

    network: Network<ContractType>;

    abstract githubRepo: string;

    constructor (
        network: Network<ContractType>,
        metadata: ProjectMetadata
    ) {
        this.network = network;
        this.metadata = metadata;
    }

    getInstance (aliasOrAddress: string) {
        if (ethers.utils.isAddress(aliasOrAddress)) {
            return this.getInstanceByAddress(aliasOrAddress);
        }
        return this.getInstanceByAlias(aliasOrAddress);
    }

    async downloadAbiFile (version: string) {
        const response = await axios.get(this.getAbiUrl(version));
        return response.data as SkaleABIFile;
    }

    getAbiUrl (version: string) {
        return `${this.githubRepo}releases/download/` +
            `${version}/${this.getAbiFilename(version)}`;
    }

    abstract getAbiFilename(version: string): string;

    getInstanceDataUrl (alias: string) {
        if (this.network instanceof ListedNetwork) {
            return `${REPOSITORY_URL}${this.network.path}/` +
                `${this.metadata.path}/${alias}.json`;
        }
        throw new NetworkNotFoundError("Network is unknown");
    }

    abstract createInstance(address: MainContractAddress):
        Instance<ContractType>;

    // Private

    private getInstanceByAddress (address: string) {
        return this.createInstance(address);
    }

    private async getInstanceByAlias (alias: string) {
        const response = await axios.get(this.getInstanceDataUrl(alias));
        if (response.status === HttpStatusCode.Ok) {
            const [address] = Object.values(response.data as InstanceData);
            return this.createInstance(address);
        }
        throw new InstanceNotFound(`Can't download data for instance ${alias}`);
    }
}

File: python/src/project.py
from abc import ABC, abstractmethod
from attr import dataclass
import requests

from constants import REPOSITORY_URL
from instance import Instance, InstanceData


@dataclass
class ProjectMetadata:
    name: str
    path: str


class Project(ABC):
    def __init__(self, network, metadata: ProjectMetadata) -> None:
        super().__init__()
        self.network = network
        self._metadata = metadata

    def get_instance(self, aliasOrAddress: str):
        if self.network.w3.is_address(aliasOrAddress):
            address = aliasOrAddress
            return self.createInstance(address)
        else:
            alias = aliasOrAddress
            url = self.getInstanceDataUrl(alias)
            response = requests.get(url)
            if (response.status != 200):
                raise ValueError(f"Can't download data for instance {alias}")
            else:
                data = InstanceData.from_json(response.text)
                if not alias in data.data:
                    raise ValueError(f'Error during parsing data for {alias}')
                return self.createInstance(data.data[alias])

    def downloadAbiFile(self, version: str):
        response = requests.get(self.get_abi_url(version))
        return response.text

    def get_abi_url(self, version: str):
        return f'{self.githubRepo}releases/download/{version}/{self.get_abi_filename(version)}'

    @abstractmethod
    def getAbiFilename(self, version: str) -> str:
        pass

    def get_instance_data_url(self, alias: str):
        from network import ListedNetwork
        if isinstance(self.network, ListedNetwork):
            return f'{REPOSITORY_URL}{self.network.path}/{self._metadata.path}/{alias}.json'
        else:
            raise ValueError('Network is unknown')

    @abstractmethod
    def create_instance(self, address: str) -> Instance:
        pass
